pyro/contrib/funsor/infer/traceenum_elbo.py

```diff
@@ -0,0 +1,99 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import contextlib
+
+import funsor
+
+from pyro.distributions.util import copy_docs_from
+from pyro.infer import ELBO
+from pyro.infer import TraceEnum_ELBO as OrigTraceEnum_ELBO
+
+from pyro.contrib.funsor import to_data, to_funsor
+from pyro.contrib.funsor.handlers import enum, plate, replay, trace
+
+
+def terms_from_trace(tr):
+    """Helper function to extract elbo components from execution traces."""
+    # data structure containing densities, measures, scales, and identification
+    # of free variables as either product (plate) variables or sum (measure) variables
+    terms = {"log_factors": [], "log_measures": [], "scale": to_funsor(1.),
+             "plate_vars": frozenset(), "measure_vars": frozenset()}
+    for name, node in tr.nodes.items():
+        if node["type"] != "sample" or type(node["fn"]).__name__ == "_Subsample":
+            continue
+        # grab plate dimensions from the cond_indep_stack
+        terms["plate_vars"] |= frozenset(f.name for f in node["cond_indep_stack"] if f.vectorized)
+        # grab the log-measure, found only at sites that are not replayed or observed
+        if node["funsor"].get("log_measure", None) is not None:
+            terms["log_measures"].append(node["funsor"]["log_measure"])
+            # sum (measure) variables: the fresh non-plate variables at a site
+            terms["measure_vars"] |= frozenset(node["funsor"]["value"].inputs) | frozenset([name]) - terms["plate_vars"]
+        # grab the scale, assuming a common subsampling scale
```

Comment:
In Pyro we check for this with informative errors
https://github.com/pyro-ppl/pyro/blob/10c767f1e2ca9b2894ad84afe205eb83d346317f/pyro/infer/traceenum_elbo.py#L28-L32

What are your plans for error checking in `contrib.funsor`? I think it's fine to omit most error checking in the first pass, then we can add tests for `assert_warning` and `assert_error`.

Reply:
I'd like to try to centralize the error-checking logic and disentangle it from the elbo implementations, but beyond that I don't have specific plans yet and I'll probably leave it for the next redesign.